src/clp_ffi_py/ir/native/PySerializer.cpp

```diff
@@ -0,0 +1,576 @@
+PyType_Slot PySerializer_slots[]{
+        {Py_tp_alloc, reinterpret_cast<void*>(PyType_GenericAlloc)},
+        {Py_tp_dealloc, reinterpret_cast<void*>(PySerializer_dealloc)},
+        {Py_tp_new, reinterpret_cast<void*>(PyType_GenericNew)},
+        {Py_tp_init, reinterpret_cast<void*>(PySerializer_init)},
+        {Py_tp_methods, static_cast<void*>(PySerializer_method_table)},
+        {Py_tp_doc, const_cast<void*>(static_cast<void const*>(cPySerializerDoc))},
+        {0, nullptr}
+};
+// NOLINTEND(cppcoreguidelines-avoid-c-arrays, cppcoreguidelines-pro-type-*-cast)
+
+/**
+ * `PySerializer`'s Python type specifications.
+ */
+PyType_Spec PySerializer_type_spec{
+        "clp_ffi_py.ir.native.Serializer",
+        sizeof(PySerializer),
+        0,
+        Py_TPFLAGS_DEFAULT,
+        static_cast<PyType_Slot*>(PySerializer_slots)
+};
+
+CLP_FFI_PY_METHOD auto
+PySerializer_init(PySerializer* self, PyObject* args, PyObject* keywords) -> int {
+    static char keyword_output_stream[]{"output_stream"};
+    static char keyword_buffer_size_limit[]{"buffer_size_limit"};
+    static char* keyword_table[]{
+            static_cast<char*>(keyword_output_stream),
+            static_cast<char*>(keyword_buffer_size_limit),
+            nullptr
+    };
+
+    // If the argument parsing fails, `self` will be deallocated. We must reset all pointers to
+    // nullptr in advance, otherwise the deallocator might trigger segmentation fault.
+    self->default_init();
+
+    PyObject* output_stream{Py_None};
+    Py_ssize_t buffer_size_limit{PySerializer::cDefaultBufferSizeLimit};
+    if (false
+        == static_cast<bool>(PyArg_ParseTupleAndKeywords(
+                args,
+                keywords,
+                "O|L",
+                static_cast<char**>(keyword_table),
+                &output_stream,
+                &buffer_size_limit
+        )))
+    {
+        return -1;
+    }
+
+    // Ensure the `output_stream` has `write`, `flush`, and `close` methods
+    auto output_stream_has_method = [&](char const* method_name) -> bool {
+        PyObjectPtr<PyObject> const write_method{PyObject_GetAttrString(output_stream, method_name)
+        };
+        if (nullptr == write_method) {
+            return false;
+        }
+        if (false == static_cast<bool>(PyCallable_Check(write_method.get()))) {
+            PyErr_SetString(PyExc_TypeError, "");
+            PyErr_Format(
+                    PyExc_TypeError,
+                    "The attribute `%s` of the given output stream object is not callable.",
+                    method_name
+            );
+            return false;
+        }
+        return true;
+    };
+
+    if (false == output_stream_has_method("write")) {
+        return -1;
+    }
+    if (false == output_stream_has_method("flush")) {
+        return -1;
+    }
+    if (false == output_stream_has_method("close")) {
+        return -1;
+    }
+
+    if (0 > buffer_size_limit) {
```

Comment:
That's supposed to mean writing to the output stream for every write

Reply:
The reason to use internal buffering is to avoid calling Python level `IO[bytes].write` every time (which has overhead). This is also the behavior of normal compressor, right? Like zstd provides a configurable buffer size, so I think exposing such an API should be good. We can polish the doc string to make it more clear

Reply:
> Any write will be buffered but then immediately flushed to output stream before next write.`

btw, this is not correct. When the buffer_size_limit is 0, the buffered data will be flushed to the output stream immediately after serialization returns success, not before the next write. So technically speaking, the buffer is used as an intermediate storage, and in users' view, there's no internal buffering.